### PR TITLE
Paginate search and list items responses

### LINKED ISSUE #42: Properly implement pagination based on ADR
### Improvement

#34 added an ADR for how to handle pagination. #41 is going to update the API version to the latest for STAC. After that, we should implement the continuation token based approach in the ADR and described in the specification.

### Acceptance
Search and item listing endpoints have `next` links that follow the specification for retrieving the next page of results.

### PR DESCRIPTION
**Pull request by jisantuc**
## Overview

This PR paginates the search and item list responses based on the ADR in #39. It's WIP because it doesn't actually do that yet, just adds types so far and tests that the opaque token generation and parsing works. It also downgrades the version of openjdk that we build in to 8, since we use that in a few other places (including in GeoTrellis CI) and don't have the same flaky memory issues that this PR has in CI.

### Checklist

- [x] New tests have been added or existing tests have been modified

### Notes

~Based on description of the [context object](https://github.com/radiantearth/stac-api-spec/blob/43799ad169da69d5558f3b492665d90bb0c63f0f/extensions/context/README.md#context-object), I think I've done the right thing with the count -- since the pagination token in our case is a filter on the results, I think it makes sense to reduce the count accordingly. Reasonable people can disagree about that though.~ (like an hour later: I no longer agree with what I said previously and I'm making the count stable regardless of what page you're on -- f4bc955)

Also, the GET search endpoint refuses to encode either the query or the `intersects` filter in the url. I think that's a sensible choice but I'm not sure how to document it. 🤔 

### Testing Instructions

- bring up your server
- browse around in the API, tossing query parameters onto `/collections/<id>/items?...` and `/search`, making sure that the next link cooperates
- create this nice json object, subbing in a reasonable collection id, in `filters.json`

```json
{
    "collections": ["a-collection-id-in-your-db"],
    "limit": 1
}
```

- `cat filters.json | http :9090/search` -- inspect the link that you get back and make sure that its extra fields make sense based on what's in the [paging extension](https://github.com/radiantearth/stac-api-spec/blob/7a6e12868113c94b17dead8989f49d978d5dd865/api-spec.md#paging-extension) description of what to return with posts
- provide some other parameters in `filters.json` to make sure that they make sense too

Closes #42

### CODE REVIEW

File: application/src/main/scala/com/azavea/franklin/api/endpoints/SearchEndpoints.scala [hunk 1]
```diff
@@ -39,12 +45,20 @@ object SearchEndpoints {
...
           )
         }
       )(sf => (sf.datetime, sf.bbox, Some(sf.collections), Some(sf.items), sf.limit, sf.next))
 
+  val searchPostInput: EndpointInput[SearchFilters] =
+    jsonBody[SearchFilters]
+      .and(nextToken)
+      .map[SearchFilters]({ (tup: (SearchFilters, Option[PaginationToken])) =>
+        val (filters, token) = tup
+        filters.copy(next = filters.next orElse token)
```
Comment (jisantuc) [inline application/src/main/scala/com/azavea/franklin/api/endpoints/SearchEndpoints.scala:59]: This could arguably always be replacement with the query param token -- I didn't do that because I'm counting on no one having access to the information they'd need to construct the non-opaque version and just using the information that comes back in the search result `links` field to find their way around (and that `body` always has `next` set to `null`).

File: application/src/main/scala/com/azavea/franklin/api/schemas/package.scala [hunk 2]
```diff
@@ -76,4 +77,6 @@ package object schemas {
   implicit val codecStacItem: Codec.JsonCodec[StacItem] =
     jsonCodec.mapDecode(decStacItem)(encStacItem)
 
+  implicit val codecPaginationToken: Codec.PlainCodec[PaginationToken] =
+    Codec.string.mapDecode(PaginationToken.decPaginationToken)(PaginationToken.encPaginationToken)
```
Comment (jisantuc) [inline application/src/main/scala/com/azavea/franklin/api/schemas/package.scala:81]: This felt like magic

File: application/src/main/scala/com/azavea/franklin/extensions/paging/PagingLinkExtension.scala.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      Map(
+        "method"  -> "POST".asJson,
+        "headers" -> linkExtensionFields.headers.asJson,
+        "body"    -> linkExtensionFields.body.asJson,
+        "merge"   -> linkExtensionFields.merge.asJson
+      )
+    )
+
+  implicit val linkExtensionPagingLinkExtension: LinkExtension[PagingLinkExtension] =
+    LinkExtension.instance
```
Comment (jisantuc) [inline application/src/main/scala/com/azavea/franklin/extensions/paging/PagingLinkExtension.scala.scala:35]: this also (with the `syntax._` import in `StacItemDao` felt like magic.

File: application/src/main/resources/migrations/V4__add_serial_pk.sql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+-- I confirmed that NOT NULL on a SERIAL column fills in values appropriately
```
[the remark this answers is not in the document]
Comment (jisantuc) [reply application/src/main/resources/migrations/V4__add_serial_pk.sql:1]: There currently isn't a production, so I'm not super worried about it yet, but your understanding is correct

**Comment by jisantuc** [comment]
I think the behavior is correct, but confusing -- can you verify that the response from the first request included the `method` and `body` parameters?

There's an explanation of this case in the pagination section of the STAC API spec. It begins:

> OAFeat does not support POST requests for searches, however the STAC API spec does. Hypermedia links are not designed for anything other than GET requests, so providing a next link for a POST search request becomes problematic.

tl;dr: once you're POSTing, you always have to POST (and the `method` field of the `next` link should tell you that you have to POST). The `body` in the `next` link also includes your limit.

**Comment by jisantuc** [comment]
The `next` link existing when there are no more results will occur only when you get to the last page and the limit evenly divides the number of results. That has a 1/n probability of occurring. That doesn't seem too bad to me. The alternative is an extra check to see whether there are any more results. Often that would be pretty costly for deep pages, but that doesn't apply here because of our super fancy pagination strategy. I'm talking myself into this as I type more words. I'll make that change.

**Comment by jisantuc** [comment]
@CloudNiner f97424c ensure that no `next` link is returned unless there are more results available (re-using the `next` logic from the rest of the request -- this PR has been a blast)

File: application/src/main/resources/migrations/V4__add_serial_pk.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (CloudNiner) [reply application/src/main/resources/migrations/V4__add_serial_pk.sql:1]: 👍 no concerns there then!

**Comment by CloudNiner** [comment]
> ensures that no next link is returned unless there are more results

Nice, glad that was do-able.

> this PR has been a blast

Yeah, it seems like it. There's more changes than I expected and its more complicated too. Works well though!

> can you verify that the response from the first request included the method and body parameters?

Here's the full `"links"` object of the response for hte first request:
```
"links": [
    {
      "method": "POST",
      "headers": {},
      "body": {
        "bbox": null,
        "datetime": null,
        "intersects": null,
        "collections": [],
        "items": [],
        "limit": 1,
        "query": {
          "layer:ids": {
            "superset": [
              "layer-us"
            ]
          }
        },
        "next": null
      },
      "merge": false,
      "href": "http://localhost:9090/search?next=eyJ0aW1lc3RhbXBBdExlYXN0IjoiMjAyMC0wNi0xNVQxNTo1Mjo1MS43OTE1NTRaIiwic2VyaWFsSWRHcmVhdGVyVGhhbiI6MX0=",
      "rel": "next",
      "type": "application/json",
      "title": null
    }
  ]
```
So yeah, method and body are there and correctly populated, `next` field is null.

> once you're POSTing, you always have to POST (and the method field of the next link should tell you that you have to POST). The body in the next link also includes your limit.

Right, makes sense. So in this case I would post again with the body returned in the links section of the first result. If I do that, I get the second entry in the collection, which is correct 👍
